### PR TITLE
 Handle C# "for" statement in control flow graph

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

**Comment by AlekseyTs** [comment]
@mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

**Review by mavasani — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/ForLoopBinder.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/ForLoopBinder.cs]: Perhaps move `if` inside `new BoundBlock`:
```
increment = new BoundBlock(
    increment.Syntax,
    locals,
    increment.Kind == BoundKind.StatementList ?
        ((BoundStatementList)increment).Statements :
        ImmutableArray.Create(increment))
    { WasCompilerGenerated = true };
```

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Need a second review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Need a second review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Need a second review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please, need a second review.

**Comment by jcouv** [comment]
Sorry for the delay. Looking now.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt
Comment (jcouv) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: Not related to this PR: I don't understand why this file always gets a lot of diff noise.
FYI, I filed [https://github.com/dotnet/roslyn/issues/26285](https://github.com/dotnet/roslyn/issues/26285)

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: Perhaps add xml doc

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Nit: extra empty line below

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs]: Would it be interesting to test a conditional expression with a throw case?

**Review by jcouv — APPROVED** [review]
LGTM Thanks

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs]: > Would it be interesting to test a conditional expression with a throw case?

I think not particularly. The purpose of this test is to confirm that condition goes through appropriate conditional branch rewriting. We don't need to test all possible constructs that can cause control flow inside it. Throw expressions are covered by other existing targeted tests.

---
In reply to: [183110773](https://github.com/dotnet/roslyn/pull/26202#discussion_r183110773) [](ancestors = 183110773)

**Pull request merged**
